lib/block/merkleblock.js

```diff
@@ -0,0 +1,285 @@
+ */
+MerkleBlock.prototype.toJSON = function toJSON() {
+  return JSON.stringify(this.toObject());
+};
+
+/**
+ * Verify that the MerkleBlock is valid
+ * @returns {Bool} - True/False whether this MerkleBlock is Valid
+ */
+MerkleBlock.prototype.validMerkleTree = function validMerkleTree() {
+  $.checkState(this.flags instanceof Array, 'MerkleBlock flags is not an array');
+  $.checkState(this.hashes instanceof Array, 'MerkleBlock flags is not an array');
+  var self = this;
+  if(this._validMerkleTree === true) {
+    return true;
+  } else if (this._validMerkleTree === false) {
+    return false;
+  }
+
+  // Can't have more hashes than numTransactions
+  // TODO: Test for this condition
+  if(this.hashes.length.length > this.numTransactions) {
+    return this._setValidMerkleTree(false);
+  }
+
+  // Can't have more flag bits than num hashes
+  // TODO: Test for this condition
+  if(this.flags.length * 8 < this.hashes.length) {
+    return this._setValidMerkleTree(false);
+  }
+
+  // Calculate height of tree
+  // From Bitcoin Core merkleblock.h CalcTreeWidth() + CPartialMerkleTree
+  var height = 0;
+  while (calcTreeWidth(height) > 1) {
+    height++;
+  }
+
+  var txs = [];
+  var flagBitsUsed = 0;
+  var hashesUsed = 0;
+
+  var root = traverse(height, 0);
+  if(hashesUsed !== this.hashes.length) {
+    return this._setValidMerkleTree(false);
+  }
+  return this._setValidMerkleTree(BufferUtil.equals(root, this.header.merkleRoot));
+
+  // Modeled after Bitcoin Core merkleblock.cpp TraverseAndExtract()
+  function traverse(depth, pos) {
+    if(flagBitsUsed > self.flags.length * 8) {
+      return null;
+    }
+    var isParentOfMatch = (self.flags[flagBitsUsed >> 3] >>> (flagBitsUsed++ & 7)) & 1;
+    if(depth === 0 || !isParentOfMatch) {
+      if(hashesUsed >= self.hashes.length) {
+        return null;
+      }
+      var hash = self.hashes[hashesUsed++];
+      if (depth === 0 && isParentOfMatch) {
+        txs.push(hash)
+      }
+      return new Buffer(hash, 'hex');
+    } else {
+      var left = traverse(depth-1, pos*2);
+      var right;
+      if(pos*2+1 < calcTreeWidth(depth-1)) {
+        right = traverse(depth-1, pos*2+1);
+      } else {
+        right = left;
+      }
+      return Hash.sha256sha256(new Buffer.concat([left, right]));
+    }
+  }
+
+  function calcTreeWidth(height) {
+    return (self.numTransactions + (1 << height) - 1) >> height;
+  }
+}
+
+/**
+ * @param {Transaction|String} - Transaction or Transaction ID Hash
+ * @returns {Bool} - return true/false if this MerkleBlock has the TX or not
+ * @private
+ */
+MerkleBlock.prototype.hasTransaction = function hasTransaction(tx) {
+  $.checkArgument(!_.isUndefined(tx), 'No transaction given');
+  $.checkArgument(tx instanceof Transaction
+    || typeof tx === 'string', 'No transaction given');
+
+  var hash = tx;
+  if(tx instanceof Transaction) {
+    hash = tx.id;
+  }
+  var revHash = BufferUtil.reverse(new Buffer(hash,'hex')).toString('hex');
```

Comment:
Wasn't sure if I should check normal + reversed version of the hash.  The reason is, the hashes in the `MerkleBlock` will be reversed, so I could default to just reversing what this method gets in always, but it may be nice to be able to pass them in either way?